theseus/core/robust_cost_function.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Tuple
+
+import torch
+
+import theseus as th
+
+from .loss import Loss
+
+
+class RobustCostFunction(th.CostFunction):
+    def __init__(
+        self,
+        loss_function: Loss,
+        log_loss_radius: th.Vector,
+        cost_function: th.CostFunction,
+    ):
+        self.loss_function = loss_function
+        self.cost_function = cost_function
+        self.log_loss_radius = log_loss_radius
+        self._optim_vars_attr_names = self.cost_function._optim_vars_attr_names
+        self._aux_vars_attr_names = self.cost_function._aux_vars_attr_names
+        self.register_aux_var("log_loss_radius")
+
+    def error(self) -> torch.Tensor:
+        return self.cost_function.error()
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.cost_function.jacobians()
+
+    def weighted_error(self) -> torch.Tensor:
+        return self.cost_function.weighted_error()
+
+    def weighted_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.weighted_jacobians_error()
+
+    def value(self) -> torch.Tensor:
+        weighted_error = self.weighted_error()
+        squared_norm = torch.sum(weighted_error**2, dim=1, keepdim=True)
+        loss_radius = torch.exp(self.log_loss_radius.data)
+        return self.loss_function.evaluate(squared_norm, loss_radius)
+
+    def rescaled_jacobians_error(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
```

Comment:
I am wondering if we can simply the API here and avoid introducing this new function.

If `weighted_jacobians_error()` is overridden for `RobustCostFunction` to do what `rescaled_jacobians_error()` is doing then you don't need a separate new function for this.

Reply:
Since this feature is not ready to merge to main and has not been fully tested, I just implemented another function for simplification.